src/mjolnir/linkclassification.cc

```diff
@@ -439,6 +423,311 @@ struct LinkGraphBuilder {
+  }
+};
+
+// A - first_node
+// E - last_node
+// AB - fork_edge
+// DE - merge_edge
+// A-F-E - link_edges(slip lane)
+//   C____D___E___
+//  |        /
+//  |      ^
+//  B     /
+//  |   F
+//  | /
+//  A
+//  |
+// The idea is to check whether edges fit "triangle" pattern, i.e
+// when there another two ways from A to E:
+// - A-F-E though the slip lane
+// - A-B-C-D-E through the intersection C
+// e.g https://www.openstreetmap.org/way/5277672
+bool IsSlipLane(Data& data, SlipLaneInput input, double traverse_threshold) {
+  // traversing from fork_edge(AB) in forward direction hoping to reach the intersection(C)
+  auto forward_nodes =
+      GoTowardsIntersection(input.first_node, input.fork_edge, true, traverse_threshold, data);
+  // traversing from merge_edge(ED) in reverse direction hoping to reach the intersection(C)
+  auto reverse_nodes =
+      GoTowardsIntersection(input.last_node, input.merge_edge, false, traverse_threshold, data);
+
+  // check if two directions intersect
+  boost::optional<uint32_t> intersection_node;
+  for (auto node : reverse_nodes) {
+    if (std::find(forward_nodes.begin(), forward_nodes.end(), node) != forward_nodes.end()) {
+      intersection_node = node;
+      break;
+    }
+  }
+
+  if (intersection_node) {
+    LOG_DEBUG("\n" + VisualizeIntersection(data,
+                                           {{forward_nodes, input.first_node, "#ff0000"},
+                                            {reverse_nodes, input.last_node, "#0000ff"}},
+                                           *intersection_node));
+  }
+
+  return intersection_node.has_value();
+}
+
+SlipLaneInput GetSlipLaneInput(Data& data, const std::vector<uint32_t>& link_edges) {
+  auto edge_heading = [&](uint32_t from_node, const Edge& edge) -> double {
+    auto shape = EdgeShape(data, edge);
+    if (edge.sourcenode_ == from_node)
+      return shape[0].Heading(shape[1]);
+    return shape[shape.size() - 1].Heading(shape[shape.size() - 2]);
+  };
+
+  // finds neighbour edge with min absolute heading delta
+  auto find_closest_neighbour_edge = [&](uint32_t node, const Edge& edge, bool forward) -> size_t {
+    double min_angle_delta = 360;
+    size_t best_edge_idx = SlipLaneInput::kInvalidEdge;
+    double origin_heading = edge_heading(node, edge);
+    auto bundle = collect_node_edges(data.nodes[node], data.nodes, data.edges);
+    for (auto to : bundle.node_edges) {
+      if (to.first.attributes.link || !IsEdgeDriveableInDirection(node, to.first, forward))
+        continue;
+
+      double neighbour_heading = edge_heading(node, to.first);
+      double abs_delta = std::abs(origin_heading - neighbour_heading);
+      double delta = std::min(abs_delta, 360 - abs_delta);
+      if (delta < min_angle_delta) {
+        min_angle_delta = delta;
+        best_edge_idx = to.second;
+      }
+    }
+    return best_edge_idx;
+  };
+
+  SlipLaneInput res;
+  Edge first_link_edge = *data.edges[link_edges.back()];
```

Comment:
it's not clear why **first** link edge is the last one from `link_edges` list. We should add a comment about that, or reverse `link_edges` list before call `GetSlipLaneInput` function